File: lib/ts/TransactionLoader.ts
import { TraceTransactionLog, TraceProject } from '../../models/ts/Tracer_pb';
import { ITransactionReader } from './ITransactionReader';

export class TransactionLoader {
    protected transactionLogCache: { [projectId: string]: { [partition: string]: TraceTransactionLog } } = {};

    constructor(protected transactionReader: ITransactionReader) {
    }

    public async LoadTraceTransactionLog(project: TraceProject, partition: string, cacheBuster: string): Promise<TraceTransactionLog> {
        let projectCache = this.transactionLogCache[project.getId()];
        if (projectCache && projectCache[partition]) {
            return projectCache[partition];
        }

        const traceTransactionLog: TraceTransactionLog = await this.transactionReader.GetTransactionLog(project, partition, cacheBuster);
        // console.log(`Loaded Transaction Log: ${JSON.stringify(traceTransactionLog.toObject())}`);

        if (!projectCache) {
            projectCache = {};
            this.transactionLogCache[project.getId()] = projectCache;
        }
        projectCache[partition] = traceTransactionLog;

        return traceTransactionLog;
    }

    public async GetTransactionLog(
        project: TraceProject,
        partition: string,
        transactionLogs: TraceTransactionLog[],
        cacheBuster: string) {
        const transactionLog = await this.LoadTraceTransactionLog(project, partition, cacheBuster);
        if (transactionLog == null) { return; }
        transactionLogs.push(transactionLog);
    }

    public async GetTransactionLogs(
        project: TraceProject,
        startTime: number,
        endTime: number,
        cacheBuster?: string): Promise<TraceTransactionLog[]> {
        const transactionLogs: TraceTransactionLog[] = new Array<TraceTransactionLog>();
        const partitions = await this.transactionReader.GetPartitionsForRange(project, startTime, endTime, cacheBuster);

        const tasks = [];
        for (const partitionKey in partitions) {
            if (!partitions.hasOwnProperty(partitionKey)) {
                continue;
            }

            const partition = partitions[partitionKey];
            tasks.push(this.GetTransactionLog(project, partition, transactionLogs, cacheBuster));
        }
        await Promise.all(tasks);

        return transactionLogs;
    }
}

File: lib/cs/TransactionLoader.cs

using System;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;
using Tracer;

public abstract class TransactionLoader
{
    protected Dictionary<string, Dictionary<int, TraceTransactionLog>> transactionLogCache = new Dictionary<string, Dictionary<int, TraceTransactionLog>>();

    public TransactionLoader()
    {
    }

    public async Task<TraceTransactionLog> LoadTraceTransactionLog(TraceProject project, int partition)
    {
        Dictionary<int, TraceTransactionLog> projectCache = null;
        if (this.transactionLogCache.ContainsKey(project.Id))
        {
            projectCache = this.transactionLogCache[project.Id];
            if (projectCache != null && projectCache[partition] != null)
            {
                return projectCache[partition];
            }
        }

        TraceTransactionLog traceTransactionLog = TraceTransactionLog.Parser.ParseFrom(await GetTransactionLogStream(project, partition));

        if (projectCache == null)
        {
            projectCache = new Dictionary<int, TraceTransactionLog>();
            this.transactionLogCache[project.Id] = projectCache;
        }
        projectCache[partition] = traceTransactionLog;

        return traceTransactionLog;
    }

    public async Task<ICollection<TraceTransactionLog>> GetTransactionLogs(TraceProject project, int startTime, int endTime)
    {
        var transactionLogs = new List<TraceTransactionLog>();
        var partitions = await GetPartitionsForRange(project, startTime, endTime);


        foreach (var partition in partitions)
        {
            if (!partitions.ContainsKey(partition.Key))
            {
                continue;
            }

            var transactionLog = await this.LoadTraceTransactionLog(project, partition.Value);
            if (transactionLog == null) { continue; }
            transactionLogs.Add(transactionLog);
        }

        return transactionLogs;
    }

    protected abstract Task<Dictionary<string, int>> GetPartitionsForRange(TraceProject project, int startTime, int endTime);
    protected abstract Task<Stream> GetTransactionLogStream(TraceProject project, int partition);
}
